cmake/git.cmake

```diff
@@ -0,0 +1,78 @@
+# SPDX-License-Identifier: Apache-2.0
+
+find_package(Git QUIET)
+
+# Usage:
+#   git_describe(<dir> <output>)
+#
+# Helper function to get a short GIT desciption associated with a directory.
+# OUTPUT is set to the output of `git describe --abbrev=12 --always` as run
+# from DIR.
+#
+function(git_describe DIR OUTPUT)
+  if(GIT_FOUND)
+    execute_process(
+      COMMAND ${GIT_EXECUTABLE} describe --abbrev=12 --always
+      WORKING_DIRECTORY                ${DIR}
+      OUTPUT_VARIABLE                  DESCRIPTION
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+      ERROR_STRIP_TRAILING_WHITESPACE
+      ERROR_VARIABLE                   stderr
+      RESULT_VARIABLE                  return_code
+    )
+    if(return_code)
+      message(STATUS "git describe failed: ${stderr}")
+    elseif(NOT "${stderr}" STREQUAL "")
+      message(STATUS "git describe warned: ${stderr}")
+    else()
+      # Save output
+      set(${OUTPUT} ${DESCRIPTION} PARENT_SCOPE)
+    endif()
+  endif()
+endfunction()
+
+# Usage:
+#   git_commit_hash(<dir> <output>)
+#
+# Helper function to extract the GIT commit hash associated with a directory.
+# OUTPUT is set to the output of `git rev-parse HEAD` as run from DIR.
+#
+function(git_commit_hash DIR OUTPUT)
+  if(GIT_FOUND)
+    execute_process(
+      COMMAND ${GIT_EXECUTABLE} rev-parse HEAD
+      WORKING_DIRECTORY                ${DIR}
+      OUTPUT_VARIABLE                  COMMIT_HASH
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+      ERROR_STRIP_TRAILING_WHITESPACE
+      ERROR_VARIABLE                   stderr
+      RESULT_VARIABLE                  return_code
+    )
+    if(return_code)
+      message(WARNING "git rev-parse failed: ${stderr}")
+    elseif(NOT "${stderr}" STREQUAL "")
+      message(WARNING "git rev-parse warned: ${stderr}")
+    else()
+      # Save output
+      set(${OUTPUT} ${COMMIT_HASH} PARENT_SCOPE)
+    endif()
+  endif()
+endfunction()
+
+# Usage:
+#   git_commit_hash_short(<full_hash> <short_hash> <short_hash_int>)
+#
+# Helper function to truncate a full GIT commit hash to a value that
+# fits in a uint32_t. SHORT_HASH is set to the first 8 characters of
+# FULL_HASH, while SHORT_HASH_INT is the decimal value equivalent of
+# 0x${SHORT_HASH}
```

Comment:
I appreciate the feedback, but specifically by calling it `git_commit_hash_short` I hope I can wave away some of the later concerns.
What does it do when the argument is shorter than 8 characters? Doesn't matter, a git commit hash is always 40 characters.
What does it do when the input is not valid hex? Doesn't matter, a git commit hash is always valid hex.

Documenting every way that using a function can go wrong is a rabbit warren with no way to predict the outputs. Just imagine the craziness that could occur is someone tried to provide the variable name `BOARD` as the second parameter.

Reply:
> Documenting every way that using a function can go wrong is a rabbit warren with no way to predict the outputs.

It depends, there's really no general answer to this question. It depends on the function's complexity, trust boundaries, safety demands, etc. https://en.wikipedia.org/wiki/Defensive_programming

With CMake the main concern is rarely safety or security (although reproducible builds matter...) but _maintenance_ because the smallest mistake can cost hours and hours of debugging, examples: d4b0273ab0c4, 13df15845e9b, 40c2e08e8211, 4cba9e6d425e, a8c0123d3c54, 774330f3fa5b, 44d9f2ce5cf1, c240b69956da

These are just the examples I personally fixed. In fact I bet it's hard to find a CMake fix that did NOT require hours and hours of debugging :-D

Anyway I didn't ask for "documentation" and why I didn't ask for "every way" either, please don't blow this out of proportion.

> Doesn't matter, a git commit hash is always 40 characters.

Errr... no: people and code truncate them all the time and they still call them "commit hash".

> Doesn't matter, a git commit hash is always valid hex.

Yes but the output of `git describe` can be many things and parsing is always buggy?

You're making the usual mistake of assuming bug-free code and it's easy to get some of these wrong. It's also very easy to get lost in this gazillion of different "versions".  No doubt your code is perfect but in a few months someone will make a "small" change and things will break in some obscure fashion. It's also very rare to find someone who understands anything to CMake code in general and even rarer to find someone who understands anything to Zephyr's extremely elaborate (and useful) CMake code (thanks for being one of these few BTW). Once you add CMake's appalling error handling and reporting and complete lack of debugging support on top of it, then I think it's only fair to set the "documentation" bar (actually: just very short comments in the source) a little bit higher and a bit more "defensively" than usual.

Reply:
Added input validation instead and manually tested arguments.